internal/mirror/mirror.go

```diff
@@ -238,94 +212,212 @@ func (s *Server) processMirror(ctx context.Context, w http.ResponseWriter, deadl
 
 		status.Saved = true
+		logger.Debugw("successfully saved mirrored archive",
+			"upstream_file", filename,
+			"local_file", writeFilename)
+
 		indexObjects = append(indexObjects, status)
 	}
 
-	// Write new index file
+	// Build and write a new index file.
 	sort.Slice(indexObjects, func(i, j int) bool {
 		return indexObjects[i].Order < indexObjects[j].Order
 	})
-	logger.Infof("index %+v", indexObjects)
-	syncFilenames := make([]*mirrordb.SyncFile, 0, len(indexObjects))
+
+	syncFilenames := make([]*mirrordatabase.SyncFile, 0, len(indexObjects))
 	filenames := make([]string, 0, len(indexObjects))
 	for _, obj := range indexObjects {
 		// Only persist state of the files we got to.
 		if obj.Saved {
-			syncFilenames = append(syncFilenames,
-				&mirrordb.SyncFile{
-					RemoteFile: obj.Filename,
-					LocalFile:  obj.LocalFilename,
-				})
-			filenames = append(filenames, fmt.Sprintf("%s/%s", mirror.FilenameRoot, obj.LocalFilename))
+			syncFilenames = append(syncFilenames, &mirrordatabase.SyncFile{
+				RemoteFile: obj.Filename,
+				LocalFile:  obj.LocalFilename,
+			})
+			filenames = append(filenames, urlJoin(mirror.FilenameRoot, obj.LocalFilename))
 		}
 	}
-	data := []byte(strings.Join(filenames, "\n"))
-	indexName := fmt.Sprintf("%s/index.txt", mirror.FilenameRoot)
+
 	bsCtx, cancel := context.WithTimeout(ctx, time.Minute)
 	defer cancel()
-	if err := blobstore.CreateObject(bsCtx, mirror.CloudStorageBucket, indexName, data, true, storage.ContentTypeTextPlain); err != nil {
-		logger.Errorw("unable to write index to blobstore", "filename", "index.txt", "error", err)
-		return nil
+
+	parent := mirror.CloudStorageBucket
+	pth := urlJoin(mirror.FilenameRoot, "index.txt")
+	data := []byte(strings.Join(filenames, "\n"))
+	if err := blobstore.CreateObject(bsCtx, parent, pth, data, true, storage.ContentTypeTextPlain); err != nil {
+		merr = multierror.Append(merr, fmt.Errorf("failed to write index.txt to blobstore: %w", err))
 	}
 
-	// sync state to DB.
+	// Save state to database.
 	if err := s.mirrorDB.SaveFiles(ctx, mirror.ID, syncFilenames); err != nil {
-		logger.Errorw("error saving index state to database", "mirrorID", mirror.ID, "error", err)
+		merr = multierror.Append(merr, fmt.Errorf("failed to save index state to database: %w", err))
+	}
+
+	retErr = merr.ErrorOrNil()
+	return
+}
+
+// computeActions computes the actions to take based on the provided known files
+// and files in the index.
+func computeActions(knownFiles []*model.MirrorFile, indexFiles []string) map[string]*FileStatus {
+	actions := make(map[string]*FileStatus, len(indexFiles))
+	for _, kf := range knownFiles {
+		actions[kf.Filename] = &FileStatus{
+			MirrorFile: kf,
+		}
+	}
+
+	for i, indexFile := range indexFiles {
+		lastIdx := strings.LastIndex(indexFile, "/")
+		fileName := indexFile[lastIdx+1:]
+
+		// Start ordering a 1 because a value of 0 is also the default value.
+		order := i + 1
```

Comment:
this is fine, but there shouldn't be any that end up in actions and don't have order set to something.